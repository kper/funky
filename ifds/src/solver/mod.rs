//! The implementation for extracting the taints from the graph.

use crate::icfg::graph::{Edge, Graph};
use anyhow::{Context, Result};
use std::collections::HashSet;

type PC = usize;

pub struct IfdsSolver;

/// Represents the taints from the execution of the solver.
#[derive(Debug)]
pub struct Taint {
    pub variable: String,
    pub function: String,
    pub pc: PC,
}

/// Request is the datastructure which defines at what instruction the [`Graph`] should be build from.
#[derive(Debug, Clone)]
pub struct Request {
    pub variable: Option<String>,
    pub function: String,
    pub pc: PC,
}

pub trait Solver {
    /// Get all taints for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Result<Vec<Taint>>;

    /// Check if the statement at `req` and the statement `resp` have a taint.
    fn is_taint(&mut self, graph: &mut Graph, req: &Request, resp: &Request) -> Result<bool>;

    /// Return all tainted variables names at any statement for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn sinks_var(&mut self, graph: &mut Graph, req: &Request) -> Result<HashSet<String>>;
}

impl Solver for IfdsSolver {
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Result<Vec<Taint>> {
        assert!(req.variable.is_some());
        assert!(req.variable.as_ref().unwrap().starts_with("%"));

        let function = &req.function;

        let taints: Vec<_> = graph
            .edges
            .iter()
            .filter(|x| {
                matches!(x, Edge::Path { .. })
                    && &x.to().function == function
                    //&& x.get_from().var_is_taut
                    && x.get_from().next_pc == req.pc
            })
            .map(|x| x.to())
            .filter(|x| !x.var_is_taut)
            .map(|x| Taint {
                function: x.function.clone(),
                pc: x.next_pc.checked_sub(1).unwrap_or(0),
                variable: x.belongs_to_var.clone(),
            })
            .collect();

        Ok(taints)
    }

    fn sinks_var(&mut self, graph: &mut Graph, req: &Request) -> Result<HashSet<String>> {
        assert!(req.variable.is_some());
        assert!(req.variable.as_ref().unwrap().starts_with("%"));

        let function = &req.function;

        let taints = graph
            .edges
            .iter()
            .filter(|x| {
                matches!(x, Edge::Path { .. })
                    && &x.to().function == function
                    && x.get_from().next_pc == req.pc
            })
            .map(|x| x.to())
            .filter(|x| !x.var_is_taut)
            .map(|x| x.belongs_to_var.clone())
            .collect::<HashSet<_>>();

        Ok(taints)
    }

    fn is_taint(&mut self, graph: &mut Graph, req: &Request, resp: &Request) -> Result<bool> {
        let all_sinks = self
            .all_sinks(graph, req)
            .context("Cannot compute the sinks")?;

        let var = resp
            .variable
            .as_ref()
            .context("Please specify a variable for the response")?;

        Ok(all_sinks
            .into_iter()
            .find(|x| x.function == resp.function && x.pc == resp.pc && &x.variable == var)
            .is_some())
    }
}

pub trait GraphReachability {
    /// Get all taints for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Vec<Taint>;
}
