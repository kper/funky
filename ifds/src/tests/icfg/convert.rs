use crate::icfg::convert::ConvertSummary;
use crate::icfg::tikz::render_to;
use crate::solver::Request;
use insta::assert_snapshot;
use log::error;

use crate::grammar::*;

use crate::icfg::flowfuncs::taint::flow::TaintNormalFlowFunction;
use crate::icfg::flowfuncs::taint::initial::TaintInitialFlowFunction;

use std::fs::{OpenOptions, create_dir};
use std::io::Write;

/// Write the IR to a seperate file. This makes it possible
/// to run it in the UI.
fn write_ir(name: &str, ir: &str) {
    let _ = create_dir("src/tests/icfg/ir_code");
    let mut fs = OpenOptions::new()
        .write(true)
        .create(true)
        .open(format!("src/tests/icfg/ir_code/{}.ir", name))
        .unwrap();
    fs.write_all(&ir.as_bytes()).unwrap();
}

macro_rules! ir {
    ($name:expr, $req:expr, $ir:expr) => {
        let mut convert = ConvertSummary::new(TaintInitialFlowFunction, TaintNormalFlowFunction);

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let res = convert.visit(&prog, &$req);

        if let Err(err) = res {
            error!("ERROR: {}", err);
            err.chain()
                .skip(1)
                .for_each(|cause| error!("because: {}", cause));
            panic!("")
        }

        write_ir($name, $ir);

        let (graph, state) = res.unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("{}_dot", $name), output);
    };
}

#[test]
fn test_ir_const() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_const",
        req,
        "
         define test (result 0) (define %0) {
            %0 = 1
         };
    "
    );
}


#[test]
fn test_ir_simple_store() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_simple_store",
        req,
        "
         define test (param %0) (result 0) (define %0 %1) {
            STORE %0 AT 0 + %0 ALIGN 2 32
         };
    "
    );
}

#[test]
fn test_ir_simple_load() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_simple_load",
        req,
        "
         define test (param %0) (result 0) (define %0 %1) {
            %1 = LOAD %0 OFFSET 0 ALIGN 0
         };
    "
    );
}

#[test]
fn test_ir_double_assign() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_double_const",
        req,
        "
         define test (result 0) (define %0 %1 %2){
            %0 = 1
            %1 = 1
            %2 = %0
            %2 = %1
         };
    "
    );
}

#[test]
fn test_ir_double_assign_with_params() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_double_const_with_params",
        req,
        "
         define test (param %0) (result 0) (define %0 %1 %2){
            %1 = 1
            %2 = %0
            %2 = %1
         };
    "
    );
}

#[test]
fn test_ir_chain_assign() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_chain_assign",
        req,
        "
         define test (result 0) (define %0 %1 %2 %3){
            %0 = 1
            %1 = 1
            %2 = %0
            %3 = %2
         };
    "
    );
}

#[test]
fn test_ir_unop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_unop",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = op %0
            %1 = op %0   
        };"
    );
}

#[test]
fn test_ir_binop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_binop",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %2 = %1 op %0   
        };"
    );
}

#[test]
fn test_ir_binop_offset() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_binop_offset",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %2 = %1 op %0   
        };"
    );
}

#[test]
fn test_ir_phi() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_phi",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = phi %0 %1
        };"
    );
}

#[test]
fn test_ir_killing_op() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_killing_op",
        req,
        "define test (result 0) (define %0 %1 %2)  {
            %0 = 1
            %1 = 1
            KILL %0
            KILL %1
            %2 = 1
        };"
    );
}

#[test]
fn test_ir_block() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_block",
        req,
        "define test (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            GOTO 1
            BLOCK 1
            %1 = 2
        };"
    );
}

#[test]
fn test_ir_if_else() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_if_else",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = 2
            %2 = 3
            GOTO 3
            BLOCK 2
            %2 = 4
            GOTO 3
            BLOCK 3
            %0 = %2
        };
        "
    );
}

#[test]
fn test_ir_if() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_if",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 0
            %1 = 2
            %2 = 3
        };
        "
    );
}

#[test]
fn test_ir_loop() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_loop",
        req,
        "define main (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            %1 = 2
            GOTO 0 
        };
        "
    );
}

#[test]
fn test_ir_table() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_table",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            %1 = 2
            %2 = 3
            BLOCK 1
            %1 = 2
            %2 = 3
            BLOCK 2
            %2 = 4
            TABLE GOTO 0 1 2 ELSE GOTO 2
        };
        "
    );
}

#[test]
fn test_ir_functions() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_functions",
        req,
        "define test (result 0) (define %0) {
            %0 = 1
            CALL mytest(%0)
        };
        define mytest (param %0) (result 0) (define %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN;
        };"
    );
}

#[test]
fn test_ir_multiple_functions() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_multiple_functions",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 <- CALL mytest(%0)
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0)  {
            RETURN %0;
        };"
    );
}

#[test]
fn test_ir_functions_rename_reg() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_functions_rename_regs",
        req,
        "define test (result 0) (define %0) {
            %0 = 1
            %0 <- CALL mytest(%0)
        };
        define mytest (param %5) (result 1) (define %5 %6)  {
            %6 = %5
            RETURN %6;
        };"
    );
}

#[test]
fn test_ir_return_values() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return_values",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %1;
        };"
    );
}

#[test]
fn test_ir_return_passed_value() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return_passed_value",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            RETURN %0;
        };"
    );
}

#[test]
fn test_ir_return_values2() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return_values2",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 2
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %1 = 3
            RETURN %0;
        };"
    );
}

#[test]
fn test_ir_return_values3() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return_values3",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 2
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            RETURN %0;
        };"
    );
}

#[test]
fn test_ir_overwrite_return_values() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_overwrite_return_values",
        req,
        "define test (result 0) (define %0) {
            %0 = 1
            %0 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %1;
        };"
    );
}

#[test]
fn test_ir_early_return() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_early_return",
        req,
        "define test (result 0) (define %0 %1 %2 %3) {
            %0 = 1
            %1 <- CALL mytest(%0)
            %2 <- CALL mytestfoo(%0)
            %3 <- CALL mytestfoo(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %1;
        };
        define mytestfoo (param %0) (result 1) (define %0 %1) {
            %1 = 3
            RETURN %0;
        };
        "
    );
}

#[test]
fn test_ir_return_double() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %0 %1 <- CALL mytest(%0)
            %1 = 2
        };
        define mytest (param %0) (result 2) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %0 %1;
        };
        "
    );
}

#[test]
fn test_ir_return_branches() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return_branches",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 5
            %1 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1 %2) {
            %1 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = 2
            %2 = 3
            RETURN %1;
            GOTO 3
            BLOCK 2
            %2 = 4
            GOTO 3
            BLOCK 3
            RETURN %0;
        };
        "
    );
}

#[test]
fn test_ir_self_loop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_self_loop",
        req,
        "define test (param %0) (result 1) (define %0 %1 %2) {
            %2 = 1
            %0 = 5
            %1 <- CALL test(%0)
            %0 = %1
            RETURN %0;
        };
        "
    );
}

#[test]
fn test_global_get_and_set() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_get_and_set",
        req,
        "
        define 0 (param %0) (result 0) (define %-2 %-1 %0 %1) {
        %1 = %-1
        %-2 = %1
        };
    "
    );
}

#[test]
fn test_global_set() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_set",
        req,
        "
        define 0 (param %0) (result 0) (define %-1 %0 %1) {
            %0 = 1
            %-1 = %0
            %1 <- CALL 1()
        };

        define 1 (param) (result 1) (define %-1 %0) {
            %0 = %-1
            RETURN %0;
        };
    "
    );
}

#[test]
fn test_global_get_and_set_multiple_functions() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_get_and_set_multiple_functions",
        req,
        "
        define 0 (param %0) (result 0) (define %-2 %-1 %0 %1 %2) {
        %1 = %-1
        %-2 = %1
        %2 = 1
        %0 <- CALL 1 (%2)
        %1 <- CALL 2 ()
        };

        define 1 (param %0) (result 1) (define %-2 %0) {
        %0 = %-2
        RETURN %0;
        };

        define 2 (result 1) (define %-2 %0) {
        %0 = 1
        %-2 = 1
        RETURN %0;
        };
    "
    );
}

#[test]
fn test_global_call() {
    let req = Request {
        variable: None,
        function: "1".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_call",
        req,
        "
        define 0 (param %0) (result 0) (define %-2 %-1 %0 %1) {
        BLOCK 0
        %1 = %0
        %-1 = %1
        RETURN ;
        };
        define 1 (param %0) (result 0) (define %-2 %-1 %0 %1 %2) {
        BLOCK 1
        %1 = 1
        CALL 0(%1)
        %2 = %0
        %-2 = %2
        RETURN ;
        };
    "
    );
}

#[test]
fn test_global_writes() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_globals_writes",
        req,
        "
        define test (result 0) (define %-1 %0 %2) {
            %0 = 1
            %-1 = %0 
            %2 <- CALL mytest()
        };
        define mytest (param) (result 1) (define %-1 %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN %-1;
        };
    "
    );
}

#[test]
fn test_global_check_order() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_globals_check_order",
        req,
        "
        define test (result 0) (define %-2 %-1 %0 %2) {
            %0 = 1
            %-1 = %0 
            %2 <- CALL mytest()
        };
        define mytest (param) (result 1) (define %-2 %-1 %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN %-1;
        };
    "
    );
}

#[test]
fn test_memory_store() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_memory_store",
        req,
        "
        define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7 %8 %9) {
        BLOCK 0
        %1 = -12345
        STORE %1 AT 0 + %0 ALIGN 2 32
        %2 = 8
        %3 = -12345
        STORE %3 AT 0 + %2 ALIGN 3 64
        %5 = 8
        %6 = -12345
        STORE %6 AT 0 + %5 ALIGN 2 32
        %7 = 8
        %8 = -12345
        STORE %8 AT 0 + %7 ALIGN 3 64
        RETURN ;
        };
    "
    );
}

#[test]
fn test_memory_load() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_memory_load",
        req,
        "
       define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
        BLOCK 0
        %0 = 8
        %1 = -12345
        STORE %1 AT 0 + %0 ALIGN 2 32
        %4 = 8
        %5 = LOAD %4 OFFSET 0 ALIGN 0
        %6 = 8
        %7 = LOAD %6 OFFSET 0 ALIGN 0
        KILL %7
        KILL %6
        RETURN ;
       }; 
    "
    );
}

#[test]
fn test_memory_load_different_functions() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_memory_load_different_functions",
        req,
        "
       define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
        BLOCK 0
        %0 = 8
        %1 = -12345
        STORE %1 AT 0 + %0 ALIGN 2 32
        %2 <- CALL 1 ()
        RETURN ;
       }; 

       define 1 (result 1) (define %0 %1) {
        %1 = 8
        %0 = LOAD %1 OFFSET 0 ALIGN 0
        RETURN %0;
       };
    "
    );
}

#[test]
fn test_memory_load_different_functions2() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_memory_load_different_functions2",
        req,
        "
       define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
        BLOCK 0
        %0 = 8
        %1 = -12345
        STORE %1 AT 0 + %0 ALIGN 2 32
        %2 <- CALL 1 ()
        STORE %1 AT 1 + %0 ALIGN 2 32
        %3 <- CALL 1 ()
        RETURN ;
       }; 

       define 1 (result 1) (define %0 %1) {
        %1 = 8
        %0 = LOAD %1 OFFSET 0 ALIGN 0
        RETURN %0;
       };
    "
    );
}


