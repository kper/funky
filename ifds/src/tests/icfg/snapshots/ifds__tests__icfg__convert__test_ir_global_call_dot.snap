---
source: src/tests/icfg/convert.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (0) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (1) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (3) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (5) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (6) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (8) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (9) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (11) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (12) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (14) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (15) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (17) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (18) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (19) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (20) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (21) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (22) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (23) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (24) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (25) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (26) at (1, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (27) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (28) at (1, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (29) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (30) at (1, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (31) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (34) at (1, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (40) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (48) at (1, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (49) at (2, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (50) at (2, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (51) at (2, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (52) at (2, 6) { };
\node[font=\tiny] (note_0) at (-1.5, 0.5) {Block(1)};
\node[font=\tiny] (note_1) at (-1.5, 1.5) {Const(1, 1.0)};
\node[font=\tiny] (note_2) at (-1.5, 2.5) {Call(0, [1], [])};
\node[font=\tiny] (note_3) at (-1.5, 3.5) {Assign(2, 0)};
\node[font=\tiny] (note_4) at (-1.5, 4.5) {Assign(-2, 2)};
\node[font=\tiny] (note_5) at (-1.5, 5.5) {Return([])};
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (2) at (7, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (4) at (9, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (7) at (9, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (10) at (9, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (13) at (9, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (16) at (9, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (32) at (9, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (33) at (8, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (35) at (7, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (36) at (7, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (37) at (7, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (38) at (7, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (39) at (7, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (41) at (9, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (42) at (10, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (43) at (8, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (44) at (9, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-1$}] (45) at (10, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (46) at (8, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (47) at (9, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (53) at (9, 0) { };
\node[font=\tiny] (note_6) at (5.5, 0.5) {Block(0)};
\node[font=\tiny] (note_7) at (5.5, 1.5) {Assign(1, 0)};
\node[font=\tiny] (note_8) at (5.5, 2.5) {Assign(-1, 1)};
\node[font=\tiny] (note_9) at (5.5, 3.5) {Return([])};
\node[font=\tiny] (note_10) at (5.5, 0.5) {Block(0)};
\node[font=\tiny] (note_11) at (5.5, 1.5) {Assign(1, 0)};
\node[font=\tiny] (note_12) at (5.5, 2.5) {Assign(-1, 1)};
\node[font=\tiny] (note_13) at (5.5, 3.5) {Return([])};
\node[font=\tiny] (note_14) at (5.5, 0.5) {Block(0)};
\node[font=\tiny] (note_15) at (5.5, 1.5) {Assign(1, 0)};
\node[font=\tiny] (note_16) at (5.5, 2.5) {Assign(-1, 1)};
\node[font=\tiny] (note_17) at (5.5, 3.5) {Return([])};
		\path[->, blue] (0) [loop right] edge  node { } (0);
		\path[->, blue] (0) [bend right] edge  node { } (1);
		\path[->, blue] (0) [bend right] edge  node { } (19);
		\path[->, blue] (0) [bend right] edge  node { } (20);
		\path[->, blue] (0) [bend right] edge  node { } (21);
		\path[->, blue] (0) [bend right] edge  node { } (22);
		\path[->, blue] (0) [bend right] edge  node { } (23);
		\path[->, blue] (1) [loop right] edge  node { } (1);
		\path[->, blue] (1) [bend right] edge  node { } (19);
		\path[->, blue] (1) [bend right] edge  node { } (26);
		\path[->, blue] (2) [loop right] edge  node { } (2);
		\path[->, blue] (2) [bend right] edge  node { } (36);
		\path[->, blue] (2) [bend right] edge  node { } (37);
		\path[->, blue] (2) [bend right] edge  node { } (38);
		\path[->, blue] (2) [bend right] edge  node { } (39);
		\path[->, blue] (4) [loop right] edge  node { } (4);
		\path[->, blue] (0) [bend right] edge  node { } (48);
		\path[->, blue] (1) [bend right] edge  node { } (20);
		\path[->, blue] (4) [bend right] edge  node { } (32);
		\path[->, blue] (0) [bend right] edge  node { } (28);
		\path[->, blue] (1) [bend right] edge  node { } (21);
		\path[->, blue] (4) [bend right] edge  node { } (33);
		\path[->, blue] (4) [bend right] edge  node { } (41);
		\path[->, blue] (0) [bend right] edge  node { } (34);
		\path[->, blue] (1) [bend right] edge  node { } (22);
		\path[->, blue] (4) [bend right] edge  node { } (42);
		\path[->, blue] (4) [bend right] edge  node { } (43);
		\path[->, blue] (4) [bend right] edge  node { } (44);
		\path[->, blue] (0) [bend right] edge  node { } (30);
		\path[->, blue] (1) [bend right] edge  node { } (23);
		\path[->, blue] (4) [bend right] edge  node { } (45);
		\path[->, blue] (4) [bend right] edge  node { } (46);
		\path[->, blue] (4) [bend right] edge  node { } (47);
		\path[->, blue] (1) [bend right] edge  node { } (49);
		\path[->, blue] (1) [bend right] edge  node { } (50);
		\path[->, blue] (1) [bend right] edge  node { } (51);
		\path[->, blue] (1) [bend right] edge  node { } (52);


                \end{tikzpicture}
            \end{document}
